controllers/clusterwidenetworkpolicy_controller.go

```diff
@@ -18,56 +18,183 @@ package controllers
+	rulesReconcileInterval = 30 * time.Second
 )
 
 // ClusterwideNetworkPolicyReconciler reconciles a ClusterwideNetworkPolicy object
 // +kubebuilder:rbac:groups=metal-stack.io,resources=events,verbs=create;patch
 type ClusterwideNetworkPolicyReconciler struct {
 	client.Client
-	Log      logr.Logger
-	Scheme   *runtime.Scheme
-	recorder record.EventRecorder
+	Log         logr.Logger
+	Scheme      *runtime.Scheme
+	Cache       *dns.DNSCache
+	policySpecs map[string]firewallv1.PolicySpec
 }
 
 // Reconcile ClusterwideNetworkPolicy and creates nftables rules accordingly
 // +kubebuilder:rbac:groups=metal-stack.io,resources=clusterwidenetworkpolicies,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=metal-stack.io,resources=clusterwidenetworkpolicies/status,verbs=get;update;patch
 func (r *ClusterwideNetworkPolicyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
+	log := r.Log.WithValues("cwnp", req.Name)
+
+	var cnwps firewallv1.ClusterwideNetworkPolicyList
+	if err := r.List(ctx, &cnwps, client.InNamespace(firewallv1.ClusterwideNetworkPolicyNamespace)); err != nil {
+		return done, err
+	}
+
+	// Check if new sets added or CNWP specs updated
+	if r.isSpecsChanged(cnwps) || r.nftableSetsAdded() {
+		return r.reconcileRules(ctx, log, cnwps)
+	}
+
+	return done, nil
+}
 
-	var clusterNP firewallv1.ClusterwideNetworkPolicy
-	if err := r.Get(ctx, req.NamespacedName, &clusterNP); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+func (r *ClusterwideNetworkPolicyReconciler) reconcileRules(ctx context.Context, log logr.Logger, cnwps firewallv1.ClusterwideNetworkPolicyList) (ctrl.Result, error) {
+	var f firewallv1.Firewall
+	nn := types.NamespacedName{
+		Name:      firewallName,
+		Namespace: firewallv1.ClusterwideNetworkPolicyNamespace,
 	}
+	if err := r.Get(ctx, nn, &f); err != nil {
+		if apierrors.IsNotFound(err) {
+			return done, err
+		}
 
-	// if network policy does not belong to the namespace where clusterwide network policies are stored:
-	// update status with error message
-	if req.Namespace != firewallv1.ClusterwideNetworkPolicyNamespace {
-		r.recorder.Event(&clusterNP, "Warning", "Unapplicable", fmt.Sprintf("cluster wide network policies must be defined in namespace %s otherwise they won't take effect", firewallv1.ClusterwideNetworkPolicyNamespace))
-		return ctrl.Result{}, nil
+		return done, client.IgnoreNotFound(err)
 	}
 
-	err := clusterNP.Spec.Validate()
-	if err != nil {
-		r.recorder.Event(&clusterNP, "Warning", "Unapplicable", fmt.Sprintf("cluster wide network policy is not valid: %v", err))
-		return ctrl.Result{}, nil
+	var services v1.ServiceList
+	if err := r.List(ctx, &services); err != nil {
+		return done, err
 	}
 
-	return ctrl.Result{}, nil
+	nftablesFirewall := nftables.NewFirewall(&cnwps, &services, f.Spec, r.Cache, log)
+	if err := nftablesFirewall.Reconcile(); err != nil {
+		return done, err
+	}
+
+	r.updateStatus(ctx, cnwps)
```

Comment:
Is the DNS cache snapshotted / persisted from time to time to the CWNPs status?

Reply:
Nope) Actually, info about sets is persisted in `PolicySpec.EgressRule.ToFQDN.Sets`. It's more convenient, in `firewallRenderer` to use/update this info from `Spec`(also, it's never cached, unlike `Status`).

Reply:
CNWP objects updated in `firewallRenderer`.